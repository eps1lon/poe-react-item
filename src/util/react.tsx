import * as React from 'react';

// text nodes dont need keys
export const CommaSeparator = () => ', ';

export const LineBreak = (key: string) => <br key={key} />;

export function intersperse(
  nodes: React.ReactNode[],
  renderSeparator: (key: string) => React.ReactNode = CommaSeparator,
) {
  if (nodes.length === 0) {
    return [];
  }

  return nodes.slice(1).reduce(
    (joined: React.ReactNode[], element, index) => {
      // push instead of concat's returnval because concat is expensive
      joined.push(renderSeparator(`separator-${index}`), element);
      return joined;
    },
    [nodes[0]],
  );
}
