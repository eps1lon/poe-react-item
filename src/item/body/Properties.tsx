import * as React from 'react';

import Property, { Props as PropertyProps } from './Property';
import { DisplayPropertyType } from './PropertyValue';
import {
  WeaponProperies,
  ArmourProperties,
  NoProperties,
  ShieldProperties,
  AugmentableValue,
} from '../poe';
import { round } from '../../util/number';

export interface Props {
  properties:
    | ArmourProperties
    | ShieldProperties
    | WeaponProperies
    | NoProperties;
}

export default class Properties extends React.PureComponent<Props> {
  public render() {
    return this.itemProperties().map(property => <Property {...property} />);
  }

  private itemProperties(): PropertyProps[] {
    const { properties } = this.props;
    switch (properties.kind) {
      case 'armour':
        return this.armourProperties(properties);
      case 'weapon':
        return this.weaponProperties(properties);
      case 'none':
        return [];
      default:
        // while the exhaustiveness check is nice for internal use
        // it's not guarenteed at runtime since this component can be consumed
        // by normal js
        // @ts-ignore
        throw new Error(`unrecognized property kind '${properties.kind}'`);
    }
  }

  private weaponProperties(props: WeaponProperies): PropertyProps[] {
    const displayed: PropertyProps[] = [];

    return displayed;
  }

  private armourProperties(
    props: ArmourProperties | ShieldProperties,
  ): PropertyProps[] {
    const displayed: PropertyProps[] = [];

    const defences: Array<['armour' | 'energy_shield' | 'evasion', string]> = [
      ['armour', 'armour'],
      ['energy_shield', 'energy shield'],
      ['evasion', 'evasion'],
    ];

    defences.forEach(([key, human]) => {
      const prop = props[key];
      if (prop !== undefined && prop.value > 0) {
        displayed.push({
          human,
          type: prop.augmented
            ? DisplayPropertyType.augmented
            : DisplayPropertyType.default,
          value: String(round(prop.value)),
        });
      }
    });

    if (
      'block' in props &&
      props.block !== undefined &&
      props.block.value > 0
    ) {
      displayed.push({
        human: 'chance to block',
        type: props.block.augmented
          ? DisplayPropertyType.augmented
          : DisplayPropertyType.default,
        value: `${round(props.block.value)}%`,
      });
    }

    return displayed;
  }
}
