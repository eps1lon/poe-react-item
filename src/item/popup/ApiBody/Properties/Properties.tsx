import * as React from 'react';

import LineContent from './LineContent';

export interface Props {
  properties: LineContent['props'][];
}

export default class Lines extends React.PureComponent<Props> {
  render() {
    const { properties } = this.props;
    return this.props.properties
      .sort((a, b) => (a.type || 0) - (b.type || 0))
      .map((line, index) => (
        <div className="displayProperty">
          <span className="lc">
            <LineContent key={index} {...line} />
          </span>
        </div>
      ));
  }
}
