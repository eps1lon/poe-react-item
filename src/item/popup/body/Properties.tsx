import * as React from 'react';

import Property, { Props as PropertyProps } from './Property';
import PropertyValue, { DisplayPropertyType } from './PropertyValue';
import {
  WeaponProperties,
  ArmourProperties,
  NoProperties,
  ShieldProperties,
} from '../../poe';
import { round, msToPerSecond, asPercentString } from '../../../util/number';
import {
  AugmentableValue,
  augmentableNotZero,
  toString as valueToString,
  isZero,
} from '../../../util/value';

export interface Props {
  properties:
    | ArmourProperties
    | ShieldProperties
    | WeaponProperties
    | NoProperties;
}

export default class Properties extends React.PureComponent<Props> {
  public static hasAny(properties: Props['properties']) {
    if (properties.quality !== undefined && properties.quality > 0) {
      return true;
    }

    switch (properties.kind) {
      case 'armour':
        const { armour, energy_shield, evasion } = properties;
        return (
          augmentableNotZero(armour) ||
          augmentableNotZero(energy_shield) ||
          augmentableNotZero(evasion)
        );
      case 'weapon':
        // at least display weapon type
        return true;
      case 'none':
      case undefined:
        return false;
      default:
        // while the exhaustiveness check is nice for internal use
        // it's not guarenteed at runtime since this component can be consumed
        // by normal js
        // @ts-ignore
        throw new Error(`unrecognized property kind '${properties.kind}'`);
    }
  }

  public render() {
    return this.itemProperties();
  }

  private itemProperties(): JSX.Element[] {
    const { properties } = this.props;

    const display_properties = [];

    if (properties.quality !== undefined && properties.quality > 0) {
      display_properties.push(
        <Property key="quality" human="quality">
          <PropertyValue
            type={DisplayPropertyType.augmented}
            value={`${properties.quality}%`}
          />
        </Property>,
      );
    }

    switch (properties.kind) {
      case 'armour':
        display_properties.push(...this.armourProperties(properties));
        break;
      case 'weapon':
        display_properties.push(...this.weaponProperties(properties));
        break;
      case 'none':
      case undefined:
        break;
      default:
        // while the exhaustiveness check is nice for internal use
        // it's not guarenteed at runtime since this component can be consumed
        // by normal js
        // @ts-ignore
        throw new Error(`unrecognized property kind '${properties.kind}'`);
    }

    return display_properties;
  }

  private weaponProperties(props: WeaponProperties): JSX.Element[] {
    const displayed: JSX.Element[] = [];

    if (augmentableNotZero(props.physical_damage)) {
      displayed.push(
        <Property key="phys" human="physical damage">
          <PropertyValue
            type={
              props.physical_damage.augmented
                ? DisplayPropertyType.augmented
                : DisplayPropertyType.default
            }
            value={`${valueToString(props.physical_damage.value)}`}
          />
        </Property>,
      );
    }

    const elementals: JSX.Element[] = [];
    // cold
    if (props.cold_damage !== undefined && !isZero(props.cold_damage)) {
      elementals.push(
        <PropertyValue
          key="cold"
          type={DisplayPropertyType.cold_damage}
          value={`${valueToString(props.cold_damage)}`}
        />,
      );
    }
    // fire
    if (props.fire_damage !== undefined && !isZero(props.fire_damage)) {
      elementals.push(
        <PropertyValue
          key="fire"
          type={DisplayPropertyType.fire_damage}
          value={`${valueToString(props.fire_damage)}`}
        />,
      );
    }
    // lightning
    if (
      props.lightning_damage !== undefined &&
      !isZero(props.lightning_damage)
    ) {
      elementals.push(
        <PropertyValue
          key="lightning"
          type={DisplayPropertyType.lightning_damage}
          value={`${valueToString(props.lightning_damage)}`}
        />,
      );
    }
    // elementals
    if (elementals.length > 0) {
      displayed.push(
        <Property key="elemental" human="elemental damage">
          {elementals}
        </Property>,
      );
    }

    if (props.chaos_damage !== undefined && !isZero(props.chaos_damage)) {
      displayed.push(
        <Property key="chaos" human="chaos damage">
          <PropertyValue
            type={DisplayPropertyType.chaos_damage}
            value={`${valueToString(props.chaos_damage)}`}
          />
        </Property>,
      );
    }

    if (augmentableNotZero(props.crit)) {
      displayed.push(
        <Property key="crit" human="critical strike chance">
          <PropertyValue
            type={
              props.crit.augmented
                ? DisplayPropertyType.augmented
                : DisplayPropertyType.default
            }
            value={`${asPercentString(props.crit.value, 2)}%`}
          />
        </Property>,
      );
    }

    if (augmentableNotZero(props.attack_time)) {
      displayed.push(
        <Property key="aps" human="attacks per second">
          <PropertyValue
            type={
              props.attack_time.augmented
                ? DisplayPropertyType.augmented
                : DisplayPropertyType.default
            }
            value={`${msToPerSecond(props.attack_time.value, 2)}%`}
          />
        </Property>,
      );
    }

    return displayed;
  }

  private armourProperties(
    props: ArmourProperties | ShieldProperties,
  ): JSX.Element[] {
    const displayed: JSX.Element[] = [];

    const defences: Array<['armour' | 'energy_shield' | 'evasion', string]> = [
      ['armour', 'armour'],
      ['energy_shield', 'energy shield'],
      ['evasion', 'evasion'],
    ];

    defences.forEach(([key, human]) => {
      const prop = props[key];

      if (augmentableNotZero(prop)) {
        displayed.push(
          <Property key={key} human={human}>
            <PropertyValue
              type={
                prop.augmented
                  ? DisplayPropertyType.augmented
                  : DisplayPropertyType.default
              }
              value={String(round(prop.value))}
            />
          </Property>,
        );
      }
    });

    if ('block' in props && augmentableNotZero(props.block)) {
      displayed.push(
        <Property key="block" human="chance to block">
          <PropertyValue
            type={
              props.block.augmented
                ? DisplayPropertyType.augmented
                : DisplayPropertyType.default
            }
            value={`${round(props.block.value)}%`}
          />
        </Property>,
      );
    }

    return displayed;
  }
}
