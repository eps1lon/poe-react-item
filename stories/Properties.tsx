import classnames from 'classnames';
import * as React from 'react';
import { storiesOf } from '@storybook/react';

import Properties, {
  Props as PropertiesProps,
} from '../src/item/body/Properties';

export interface Props extends PropertiesProps {
  rarity: 'normal' | 'magic' | 'rare';
}

const Story: React.SFC<Props> = ({ rarity, ...props }) => {
  return (
    <div className={classnames('poe-item', rarity)}>
      <section className="body">
        <Properties {...props} />
      </section>
    </div>
  );
};

storiesOf('properties', module)
  .add('shield armour', () => (
    <Story
      rarity="rare"
      properties={{ kind: 'armour', armour: { value: 1 }, block: { value: 3 } }}
    />
  ))
  .add('sacrifical garb', () => (
    <Story
      rarity="rare"
      properties={{
        quality: 5,
        kind: 'armour',
        armour: { value: 100, augmented: true },
        energy_shield: { value: 10, augmented: true },
        evasion: { value: 50.5, augmented: true },
      }}
    />
  ))
  .add('dagger', () => (
    <Story
      rarity="rare"
      properties={{
        quality: 5,
        kind: 'weapon',
        physical_damage: {
          value: 6,
          augmented: true,
        },
        cold_damage: [10, 13],
        fire_damage: 4,
        lightning_damage: [1, 400],
        attack_time: {
          value: 750,
        },
        crit: {
          value: 750,
          augmented: true,
        },
      }}
    />
  ))
  .add('claw', () => (
    <Story
      rarity="rare"
      properties={{
        kind: 'weapon',
        cold_damage: [10, 13],
        chaos_damage: [500, 600],
        attack_time: {
          value: 750,
        },
        crit: {
          value: 750,
          augmented: true,
        },
      }}
    />
  ));
