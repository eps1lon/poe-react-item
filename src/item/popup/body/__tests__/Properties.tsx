import * as React from 'react';
import * as renderer from 'react-test-renderer';

import Properties from '../Properties';

it('should display shields', () => {
  expect(
    renderer
      .create(
        <Properties
          properties={{
            armour: { value: 1 },
            block: { value: 3 },
          }}
        />,
      )
      .toJSON(),
  ).toMatchSnapshot();
});

it('should display sacrifical garb', () => {
  expect(
    renderer
      .create(
        <Properties
          properties={{
            quality: 5,
            armour: { value: 100, augmented: true },
            energy_shield: { value: 10, augmented: true },
            evasion: { value: 50.5, augmented: true },
          }}
        />,
      )
      .toJSON(),
  ).toMatchSnapshot();
});

it('should display dagger', () => {
  expect(
    renderer
      .create(
        <Properties
          properties={{
            quality: 5,
            physical_damage: {
              value: 6,
              augmented: true,
            },
            cold_damage: [10, 13],
            fire_damage: 4,
            lightning_damage: [1, 400],
            attack_time: {
              value: 750,
            },
            crit: {
              value: 750,
              augmented: true,
            },
          }}
        />,
      )
      .toJSON(),
  ).toMatchSnapshot();
});

it('should display claw', () => {
  expect(
    renderer
      .create(
        <Properties
          properties={{
            cold_damage: [10, 13],
            chaos_damage: [500, 600],
            attack_time: {
              value: 750,
            },
            crit: {
              value: 750,
              augmented: true,
            },
          }}
        />,
      )
      .toJSON(),
  ).toMatchSnapshot();
});
