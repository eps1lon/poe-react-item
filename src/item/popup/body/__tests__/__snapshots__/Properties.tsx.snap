// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display claw 1`] = `
Array [
  <div
    className="display-property"
  >
    <span>
      elemental damage
    </span>
    : 
    <span
      className="value cold_damage"
    >
      10-13
    </span>
  </div>,
  <div
    className="display-property"
  >
    <span>
      chaos damage
    </span>
    : 
    <span
      className="value chaos_damage"
    >
      500-600
    </span>
  </div>,
  <div
    className="display-property"
  >
    <span>
      critical strike chance
    </span>
    : 
    <span
      className="value augmented"
    >
      7.50%
    </span>
  </div>,
  <div
    className="display-property"
  >
    <span>
      attacks per second
    </span>
    : 
    <span
      className="value default"
    >
      1.33%
    </span>
  </div>,
]
`;

exports[`should display dagger 1`] = `
Array [
  <div
    className="display-property"
  >
    <span>
      quality
    </span>
    : 
    <span
      className="value augmented"
    >
      5%
    </span>
  </div>,
  <div
    className="display-property"
  >
    <span>
      physical damage
    </span>
    : 
    <span
      className="value augmented"
    >
      6
    </span>
  </div>,
  <div
    className="display-property"
  >
    <span>
      elemental damage
    </span>
    : 
    <span
      className="value cold_damage"
    >
      10-13
    </span>
    , 
    <span
      className="value fire_damage"
    >
      4
    </span>
    , 
    <span
      className="value lightning_damage"
    >
      1-400
    </span>
  </div>,
  <div
    className="display-property"
  >
    <span>
      critical strike chance
    </span>
    : 
    <span
      className="value augmented"
    >
      7.50%
    </span>
  </div>,
  <div
    className="display-property"
  >
    <span>
      attacks per second
    </span>
    : 
    <span
      className="value default"
    >
      1.33%
    </span>
  </div>,
]
`;

exports[`should display sacrifical garb 1`] = `
Array [
  <div
    className="display-property"
  >
    <span>
      quality
    </span>
    : 
    <span
      className="value augmented"
    >
      5%
    </span>
  </div>,
  <div
    className="display-property"
  >
    <span>
      armour
    </span>
    : 
    <span
      className="value augmented"
    >
      100
    </span>
  </div>,
  <div
    className="display-property"
  >
    <span>
      energy shield
    </span>
    : 
    <span
      className="value augmented"
    >
      10
    </span>
  </div>,
  <div
    className="display-property"
  >
    <span>
      evasion
    </span>
    : 
    <span
      className="value augmented"
    >
      50
    </span>
  </div>,
]
`;

exports[`should display shields 1`] = `
Array [
  <div
    className="display-property"
  >
    <span>
      armour
    </span>
    : 
    <span
      className="value default"
    >
      1
    </span>
  </div>,
  <div
    className="display-property"
  >
    <span>
      chance to block
    </span>
    : 
    <span
      className="value default"
    >
      3%
    </span>
  </div>,
]
`;
